File: cxx/UnionFind/QuickUnion.cpp
#include<iostream>

#include "UnionFind.hpp"

using namespace UnionFind;
using namespace std;

namespace UnionFind {
    class QuickUnion : public UnionFind {
        private:
            int N;
            int* id;
        public:
            QuickUnion();
            QuickUnion(int N);
            ~QuickUnion();

            // Add Connection between two components p and q        
            void _union(int p, int q) ;
            
            // Are p and q in the same component ?
            bool _connected(int p, int q);

            // Component identifier for p (0 to N-1)
            int _find(int key);

            // number of components
            int _count();

            // Print array sample
            void printSample();
    }; // class QuickUnion

} // namespace UnionFind

QuickUnion::QuickUnion() : N(N),id(new int[N])
{
    for (int i = 0; i<N; i++) {
        this->id[i] = i;
    }
}

QuickUnion::~QuickUnion() 
{
    cout << "Object is being destructed" << endl ;
    free(this->id);
}

int QuickUnion::_find(int key) {
    cout << "Hello from find " << endl ;
    int i = key ;
    while(this->id[i]!=i) {
        i = this->id[i] ;
    }
    return i;
}

void QuickUnion::_union(int p, int q) {
    int pid = _find(p);
    int qid = _find(q);
    if   (pid==qid) return;
    else (*this).id[pid] = qid ;
}

void QuickUnion::printSample() {
    int less;

    if (this->N <10) {
        less = this->N;
    }
    else
    {
        less = 10;
    }

    for(int i=0 ; i<less; i++ ) {
        std::cout << i << " -> " << this->id[i] << std::endl ;
    }
}

File: java/UnionFind/src/main/java/unionFind/QuickUnion.java
package unionFind;

public class QuickUnion implements UnionFind{

    private int n;
    private int count;
    private int id[];

    public QuickUnion(int n) {
        this.n = n;
        this.count = 0;
        id = new int[n];
        for (int i = 0 ; i < n; i++ ) {
            id[i] = i;
        }
    }

    public void union(int p, int q) {
        int i = find(p);
        int j = find(q);
        if (i==j) return ;
        id[i] = j;
        this.count++;
    }

    public boolean connected(int p, int q) {
        return find(p) == find(q);
    }

    public int find(int i) {
        int idx = i;
        while (id[idx]!=i) {
            idx = i ;
        }
        return idx;
    }

    public int count() {
        return count;
    }

    private void validate(int p){
        if (p<0 || p>=n)
            throw new IllegalStateException("index p is not between 0 and " + (n-1));
    }
}
